### PR TITLE
Introduce automatic ABI maintenance mechanism (1/2; prepare)

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

ABI is hard to track.

#### Summary of Changes

This is the preparation pr for the next rollup of `frozen_abi`

Part of #7738

### CODE REVIEW

**Comment by ryoqun** [comment]
@mvines The CI is green! This is implementation-only split of #8012. This is still rather big pr, but could you review this?

File: sdk/build.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+extern crate rustc_version;
+use rustc_version::{version_meta, Channel};
+
+fn main() {
+    // Copied and adapted from
+    // https://github.com/Kimundi/rustc-version-rs/blob/1d692a965f4e48a8cb72e82cda953107c0d22f47/README.md#example
+    // Licensed under Apache-2.0 + MIT (compatible because we're Apache-2.0)
```
Comment (mvines) [inline sdk/build.rs:7]:
```suggestion
    // Licensed under Apache-2.0 + MIT
```
nit: let's stick to just the facts.

File: sdk/macro-frozen-abi/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+                    ::log::error!("digest error: {:#?}", result);
+                }
+                result.unwrap();
+                if ::std::env::var("SOLANA_ABI_BULK_UPDATE").is_ok() {
+                    if #expected_digest != format!("{}", hash) {
+                        #p!("sed -i -e 's/{}/{}/g' $(git grep --files-with-matches frozen_abi)", #expected_digest, hash);
+                    }
+                    ::log::warn!("Not testing the abi digest under SOLANA_ABI_BULK_UPDATE!");
+                } else {
+                  assert_eq!(#expected_digest, format!("{}", hash), "Possibly ABI changed? Confirm the diff by rerunning before and after this test failed with SOLANA_ABI_DUMP_DIR");
```
Comment (ryoqun) [inline sdk/macro-frozen-abi/src/lib.rs:330]: This is the starting point!

**Comment by ryoqun** [comment]
> @garious / @aeyakovenko - I'm almost 100% sure that ABI management is going to be a thing that both of you will rage about tripping over after it's rolled out (maybe more @aeyakovenko ;) ), so I suggest you both participate in building the shackles. Care to review?

For devx matter, I've taken extra effort to easy the burden, btw. The implementation is really advanced/complex (at least for half-year rust programmer), well worth for its own crate but it's really encapsulated, I believe.

For devx, managing abi sanely just boils down to this test assertion: https://github.com/solana-labs/solana/pull/10335/files#r433615921

Basically, this is where you'll start to being tripped over. When you encounter this failure on CI, just run the abi test twice and `diff` the dumped abi digest text file and justify nothing odd is added/removed there, then commit the new digest. I made that digest file parsable for humans.

Or just page @ryoqun or remove `frozen_abi` if it's really urgent.

As the last resort, if you dare, take the look of the internal this PR is about to add. :)

Btw, this pr stranded for several months unfortunately for more important snapshot thing and it evolved for each rebase, I'm pretty sure this mechanism can detect most of abi changes nicely and updating is one-line change for engineers.

**Comment by ryoqun** [comment]
Wow, cool. I'll merge this finally! :)

**Comment by ryoqun** [comment]
> > @garious / @aeyakovenko - I'm almost 100% sure that ABI management is going to be a thing that both of you will rage about tripping over after it's rolled out (maybe more @aeyakovenko ;) ), so I suggest you both participate in building the shackles. Care to review?
> 
> For devx matter, I've taken extra effort to easy the burden, btw. The implementation is really advanced/complex (at least for half-year rust programmer), well worth for its own crate but it's really encapsulated, I believe.
> 
> For devx, managing abi sanely just boils down to this test assertion: https://github.com/solana-labs/solana/pull/10335/files#r433615921
> 
> Basically, this is where you'll start to being tripped over. When you encounter this failure on CI, just run the abi test twice and `diff` the dumped abi digest text file and justify nothing odd is added/removed there, then commit the new digest. I made that digest file parsable for humans.
> 
> Or just page @ryoqun or remove `frozen_abi` if it's really urgent.
> 
> As the last resort, if you dare, take the look of the internal this PR is about to add. :)
> 
> Btw, this pr stranded for several months unfortunately for more important snapshot thing and it evolved for each rebase, I'm pretty sure this mechanism can detect most of abi changes nicely and updating is one-line change for engineers.

@garious @aeyakovenko Because I got +1, I've merged this preparation pr and master is green at the moment. I guess it's now the last call for actually enabling abi maintenance (or burden). For more concrete example for what's needed by engineers to maintain the abi, please take a look at #8012, which I've rebased already.

**Pull request merged**